cpp/src/components/connectivity.cu

```diff
@@ -41,190 +40,71 @@ namespace detail {
+                                                     graph.number_of_edges,
+                                                     graph.number_of_vertices,
+                                                     d_alloc,
                                                      stream);
-
-    }
-  else
-    {
-      
-      //device memory requirements: 2n^2 + 2n x sizeof(IndexT) + 1 (for flag)
-      //( n = |V|)
-      //
-      size_t n2 = 2*nrows;
-      n2 = n2*(nrows*sizeof(ByteT) + sizeof(IndexT)) + 1;
-
-      int device;
-      cudaDeviceProp prop;
+  } else {
+    //device memory requirements: 2n^2 + 2n x sizeof(IndexT) + 1 (for flag)
+    //( n = |V|)
+    //
+    size_t n2 = 2 * graph.number_of_vertices;
+    n2 = n2 * (graph.number_of_vertices * sizeof(ByteT) + sizeof(VT)) + 1;
+
+    int device;
+    cudaDeviceProp prop;
       
-      cudaGetDevice(&device);
-      cudaGetDeviceProperties(&prop, device);
+    cudaGetDevice(&device);
+    cudaGetDeviceProperties(&prop, device);
 
-      if( n2 > prop.totalGlobalMem )
-        {
-
-          CUGRAPH_FAIL("ERROR: Insufficient device memory for SCC");
-        }
-      SCC_Data<ByteT, IndexT> sccd(nrows, p_d_row_offsets, p_d_col_ind);
-      sccd.run_scc(p_d_labels);
-      
+    if( n2 > prop.totalGlobalMem ) {
+      CUGRAPH_FAIL("ERROR: Insufficient device memory for SCC");
     }
 
-  //fill the vertex indices column:
-  //
-  thrust::sequence(thrust::device, p_d_verts, p_d_verts + nrows);
-  
-  
-}
-
-/**
- * @brief Compute connected components. 
- * The weak version (for undirected graphs, only) was imported from cuML.
- * This implementation comes from [1] and solves component labeling problem in
- * parallel on CSR-indexes based upon the vertex degree and adjacency graph.
- *
- * [1] Hawick, K.A et al, 2010. "Parallel graph component labelling with GPUs and CUDA"
- * 
- * The strong version (for directed or undirected graphs) is based on: 
- * [2] Gilbert, J. et al, 2011. "Graph Algorithms in the Language of Linear Algebra"
- *
- * C = I | A | A^2 |...| A^k
- * where matrix multiplication is via semi-ring: 
- * (combine, reduce) == (&, |) (bitwise ops)
- * Then: X = C & transpose(C); and finally, apply get_labels(X);
- *
- *
- * @param graph input graph; assumed undirected for weakly CC [in]
- * @param connectivity_type CUGRAPH_WEAK or CUGRAPH_STRONG [in]
- * @param table of 2 gdf_columns: output labels and vertex indices [out]
- */
+    SCC_Data<ByteT, VT> sccd(nrows, graph.offsets, graph.indices);
+    sccd.run_scc(labels);
+  }
 }
+} //namespace detail
 
-void connected_components(Graph *graph,
-                                    cugraph_cc_t connectivity_type,
-                                    cudf::table *table)  
-{
+template <typename VT, typename ET, typename WT>
+void connected_components(experimental::GraphCSR<VT,ET,WT> const &graph,
+                          cugraph_cc_t connectivity_type,
+                          VT *labels) {
```

Comment:
And this is just a comment (and I think we'd better address this later), but is there any benefit in taking connectivity_type over having separate functions for weakly_connected_components and strongly_connected_components? The former is beneficial if WCC and SCC have a lot of common code, but WCC and SCC are two very different algorithms (at least in our implementation).

Reply:
Don't know.  I was keeping the existing structure in the API.  I'd be inclined to make them separate.  The python code literally just calls the code with the different connectivity type and the C++ just splits on an if and calls two different implementations.